File: frontend/src/services/task.service.ts
import { apiClient } from "@/lib/api-client";

export const taskService = {
  // ── Customer Endpoints ──
  getCustomerTasks: async () => {
    return apiClient('/api/v1/customer/tasks');
  },
  
  getDashboardStats: async () => {
    return apiClient('/api/v1/customer/dashboard-stats');
  },
  
  getTaskDetails: async (taskId: number) => {
    return apiClient(`/api/v1/customer/tasks/${taskId}`);
  },
  
  createTask: async (data: any) => {
    return apiClient('/api/v1/customer/tasks', {
      method: 'POST',
      body: JSON.stringify(data),
    });
  },

  getTaskBids: async (taskId: number) => {
    return apiClient(`/api/v1/customer/tasks/${taskId}/bids`);
  },

  acceptBid: async (taskId: number, bidId: number) => {
    return apiClient(`/api/v1/customer/tasks/${taskId}/bids/${bidId}/accept`, {
      method: 'POST',
    });
  },

  requestRevision: async (taskId: number, data: { revision_note: string }) => {
    return apiClient(`/api/v1/customer/tasks/${taskId}/request-revision`, {
      method: 'POST',
      body: JSON.stringify(data),
    });
  },

  addFilesToTask: async (taskId: number, files: File[]) => {
    const formData = new FormData();
    files.forEach((file) => {
      formData.append('files', file);
    });

    return apiClient(`/api/v1/customer/tasks/${taskId}/files`, {
      method: 'POST',
      body: formData,
      // Note: apiClient should handle removing Content-Type for FormData
    });
  },

  // ── Writer Endpoints ──
  getWriterTasks: async () => {
    return apiClient('/api/v1/writer/tasks');
  },

  getWriterTaskDetails: async (taskId: number) => {
    return apiClient(`/api/v1/writer/tasks/${taskId}`);
  },

  getOpenTasksForWriter: async () => {
    return apiClient('/api/v1/writer/tasks/open');
  },

  placeBid: async (taskId: number, data: { bid_amount: number; message?: string }) => {
    return apiClient(`/api/v1/writer/tasks/${taskId}/bids`, {
      method: 'POST',
      body: JSON.stringify(data),
    });
  },

  getWriterBids: async () => {
    return apiClient('/api/v1/writer/bids');
  },

  withdrawBid: async (bidId: number) => {
    return apiClient(`/api/v1/writer/bids/${bidId}`, {
      method: 'DELETE',
    });
  },

  submitTask: async (taskId: number, data: { submission_note: string; is_final: boolean }) => {
    return apiClient(`/api/v1/writer/tasks/${taskId}/submit`, {
      method: 'POST',
      body: JSON.stringify(data),
    });
  }
};

File: Mobile APP/lib/services/task_service.dart
import '../models/bid.dart';
import '../models/task.dart';
import 'api_client.dart';

class TaskService {
  static Future<List<Task>> getMyTasks() async {
    final response = await ApiClient.get('/api/v1/customer/tasks');
    final results = response['results'] as List<dynamic>;
    return results.map((json) => Task.fromJson(json)).toList();
  }

  static Future<Task> getTask(int taskId) async {
    final response = await ApiClient.get('/api/v1/customer/tasks/$taskId');
    return Task.fromJson(response['results']);
  }

  static Future<List<Bid>> getBids(int taskId) async {
    final response = await ApiClient.get(
      '/api/v1/customer/tasks/$taskId/bids',
    );
    final results = response['results'] as List<dynamic>;
    return results.map((json) => Bid.fromJson(json)).toList();
  }

  static Future<void> acceptBid(int taskId, int bidId) async {
    await ApiClient.post(
      '/api/v1/customer/tasks/$taskId/bids/$bidId/accept',
      {},
    );
  }

  static Future<void> approveTask(int taskId) async {
    await ApiClient.post('/api/v1/customer/tasks/$taskId/approve', {});
  }

  static Future<void> requestRevision(int taskId, String note) async {
    await ApiClient.postForm(
      '/api/v1/customer/tasks/$taskId/request-revision',
      {'revision_note': note},
    );
  }

  static Future<void> createTask({
    required String title,
    required String description,
    required DateTime deadline,
    double? budget,
    bool isUrgent = false,
  }) async {
    await ApiClient.post('/api/v1/customer/tasks', {
      'title': title,
      'description': description,
      'deadline': deadline.toIso8601String(),
      'budget': budget,
      'is_urgent': isUrgent,
    });
  }
}
